archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterLambdaAccessesTest.java

```diff
@@ -0,0 +1,535 @@
+                    target.target();
+                };
+            }
+        }
+
+        JavaClasses classes = new ClassFileImporter().importClasses(Target.class, Caller.class);
+        Set<JavaMethodCall> calls = classes.get(Caller.class).getMethodCallsFromSelf();
+
+        assertThat(calls).hasSize(2);
+        calls.forEach(call -> assertThatCall(call).isFrom("call").isTo(Target.class, "target"));
+    }
+
+    @Test
+    public void imports_multiple_method_calls_from_multiple_lambda() {
+        @SuppressWarnings("unused")
+        class Target {
+            Target target() {
+                return this;
+            }
+
+            Target target(String s) {
+                return this;
+            }
+        }
+
+        @SuppressWarnings("unused")
+        class Caller {
+            private Target target;
+
+            Runnable call1() {
+                return () -> target.target().target().target();
+            }
+
+            Function<String, Target> call2() {
+                return s -> target.target(s).target();
+            }
+        }
+
+        JavaClasses classes = new ClassFileImporter().importClasses(Target.class, Caller.class);
+        Set<JavaMethodCall> calls = classes.get(Caller.class).getMethodCallsFromSelf();
+        assertThat(calls).hasSize(5);
+
+        assertThat(filterOriginByName(calls, "call1"))
+                .haveAtLeast(3, access().fromOrigin(Caller.class, "call1").toTarget(Target.class, "target"))
+                .hasSize(3);
+
+        assertThat(filterOriginByName(calls, "call2"))
+                .extracting(JavaMethodCall::getTarget)
+                .areExactly(1, targetCodeUnit("target"))
+                .areExactly(1, targetCodeUnit("target", String.class))
+                .hasSize(2);
+    }
+
+    @Test
+    public void imports_multiple_method_calls_from_multiple_lambda_in_one_method() {
+        @SuppressWarnings("unused")
+        class Target {
+            Target target(String s) {
+                return this;
+            }
+
+            Target target(int i) {
+                return this;
+            }
+        }
+
+        @SuppressWarnings("unused")
+        class Caller {
+            private Target target;
+
+            Function<String, Target> call(int i) {
+                Function<Target, Target> function = t -> t.target(i);
+                return s -> target.target(s);
+            }
+        }
+
+        JavaClasses classes = new ClassFileImporter().importClasses(Target.class, Caller.class);
+        Set<JavaMethodCall> calls = classes.get(Caller.class).getMethodCallsFromSelf();
+
+        assertThat(filterOriginByName(calls, "call"))
+                .extracting(JavaMethodCall::getTarget)
+                .areExactly(1, targetCodeUnit("target", String.class))
+                .areExactly(1, targetCodeUnit("target", int.class))
+                .hasSize(2);
+    }
+
+    @Test
+    public void imports_method_call_from_lambdas_from_multiple_callers() {
```

Comment:
Isn't that basically the same as `imports_method_call_from_lambda_without_parameter`, as `Caller1` and `Caller2` do not interact with each other?

Reply:
This was to check for ambiguity, i.e. theoretically the same synthetic method could be used in both cases (since it does not depend on the context), so we wanted to make sure if there is the identical call from two classes and we replace the origin, that we end up with two correct fixed up method calls. In the end in the bytecode there are two separate methods here and there is no problem, but we still wanted to cover it.